cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class for creating custom target gatesets which can be used for compilation."""
+
+from typing import Optional, List, Hashable, TYPE_CHECKING
+import abc
+
+from cirq import ops, protocols, _import
+from cirq.protocols.decompose_protocol import DecomposeResult
+from cirq.transformers import (
+    merge_k_qubit_gates,
+    merge_single_qubit_gates,
+)
+
+drop_empty_moments = _import.LazyLoader('drop_empty_moments', globals(), 'cirq.transformers')
+drop_negligible = _import.LazyLoader('drop_negligible_operations', globals(), 'cirq.transformers')
+expand_composite = _import.LazyLoader('expand_composite', globals(), 'cirq.transformers')
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _create_transformer_with_kwargs(func: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
+    """Hack to capture additional keyword arguments to transformers while preserving mypy type."""
+
+    def transformer(
+        circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+    ) -> 'cirq.AbstractCircuit':
+        return func(circuit, context=context, **kwargs)  # type: ignore
+
+    return transformer
+
+
+class CompilationTargetGateset(ops.Gateset, metaclass=abc.ABCMeta):
+    """Abstract base class to create gatesets that can be used as targets for compilation.
+
+    An instance of this type can be passed to transformers like `cirq.convert_to_target_gateset`,
+    which can transform any given circuit to contain gates accepted by this gateset.
```

Comment:
This class seems important enough for users that it might be good to give a snippet of a simple implementation in the docstring. WDYT ?

Reply:
Concrete implementations will be added in the follow-up PR. Maybe we can then update the docstrings here to include a hyper-linked example to implementation? Something like 

"Please see `cirq.CZTargetGateset` for a concrete implementation." ?